### PR TITLE
NoiseModelFactorN - fixed-number of variables >6

### PR DESCRIPTION
**Pull request by gchenfc**
Although this doesn't have an immediate use-case, I just wanted to work on this as a de-stressing exercise.  Will not feel bad if it is rejected.

Currently there is `NoiseModelFactor1`, `NoiseModelFactor2`, ..., `NoiseModelFactor6` but nothing higher, so if you want to implement n-way factors where n is known at compile time but greater than 6, you have to derive from `NoiseModelFactor` and implement `unwhitenedError`.  This is fine, but I thought it'd be nice to have an n-way version that still has the convenient `evaluateError` syntax.

This might also be nice to use in place of existing `NoiseModelFactor1-6` classes in the future since, if the number of arguments for a given factor ever changes, it requires fewer changes (N is figured out by the compiler rather than hard-coded).

### CODE REVIEW

**Comment by varunagrawal** [comment]
Dude, I love this!

**Review by dellaert — COMMENTED** [review]
I love this. Is there a way to replace some of the earlier factors with a typedef  to this new factor? has to be 100% compatible of course.

**Comment by gchenfc** [comment]
Thanks!

> I love this. Is there a way to replace some of the earlier factors with a typedef to this new factor? has to be 100% compatible of course.

I did consider this but decided to wait what you thought first.  Glad you feel similarly :)

I think the only extra things that wouldn't be possible with the typedef would be:
1. `inline Key key1() const { return keys_[0]; }` etc.
2. `typedef VALUE1 X1;` etc.

Because these have literal numbers in their names (and macros are processed before templates so we can't use macro magic to get these names).

After fixing the CI failure, I'll draft a potential solution and see what you think.

**Comment by dellaert** [comment]
We can just add key want to key six and have it fail if N< less than six :-) There might even be template magic to only instantiate these methods up to N.

**Comment by gchenfc** [comment]
Oh true I forgot about that!

I implemented that - it's mostly done, except that the `using` syntax:
```c++
template <class VALUE1, class VALUE2, ...., class VALUE6>
using NoiseModelFactor6 = NoiseModelFactorN<VALUE1, VALUE2, VALUE3, VALUE4, VALUE5, VALUE6>;
```
is causing some serialization issues.  Here's a MWE:
https://godbolt.org/z/WjrbfWMnT
which is perplexing because if you change `using BaseAlias = Base<T>;` to `struct BaseAlias : Base<T> {};` it works, even though `BaseAlias` is still not a type without giving it template parameters.  Or even if you use `Base` which is also templated (and how it is currently in GTSAM) it still works.

I'll think about this more later and maybe ask fan or varun for help.

**Comment by dellaert** [comment]
Ping @gchenfc - I prefer you ask help from someone not Varun or Fan - both racing to deadlines.

File: gtsam/nonlinear/NonlinearFactor.h
Comment (dellaert) [inline gtsam/nonlinear/NonlinearFactor.h:471]: Indent?

File: gtsam/nonlinear/NonlinearFactor.h
Comment (dellaert) [inline gtsam/nonlinear/NonlinearFactor.h:473]: Properly document and separate from old functions, which should be deprecated and put in deprecated section near end of class.

File: gtsam/nonlinear/NonlinearFactor.h
Comment (dellaert) [inline gtsam/nonlinear/NonlinearFactor.h:404]: I strongly dislike this :-) clutters up a *key* class in GTSAM. All to have aliases X1 etc? Do we even use them anywhere?

File: gtsam/nonlinear/NonlinearFactor.h
Comment (dellaert) [inline gtsam/nonlinear/NonlinearFactor.h:310]: A great exercise in meta-programming which I’d love to see disappear, see comment below :-)

File: gtsam/nonlinear/NonlinearFactor.h
Comment (dellaert) [inline gtsam/nonlinear/NonlinearFactor.h:32]: Why do we need this stuff? Document.

File: gtsam/mainpage.dox
Comment (dellaert) [inline gtsam/mainpage.dox:19]: Seems to conflict with “deprecated” comment below.

File: gtsam/nonlinear/NonlinearFactor.h
Comment (dellaert) [inline gtsam/nonlinear/NonlinearFactor.h:593]: Can you comment on that ?

**Review by dellaert — CHANGES_REQUESTED** [review]

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h:471]: I can indent, but clang-format auto un-indents it since it's a macro/preprocessor thing (https://google.github.io/styleguide/cppguide.html#Preprocessor_Directives)

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h:404]: Another option is to just typedef them to void if they shouldn't exist.  This makes the code cleaner at the expense of less stringent compile-time checks.  For example, something like:

```c++
using X2 = std::conditional< (N > 2), VALUES[2], void >
```

but then someone might get confused if they do e.g. `NoiseModelFactorN<double, double>::X6 myVariable = 3.2` and then the compiler will give some confusing message (as opposed to just "X2 does not exist").




> Do we even use them anywhere?

I don't think it's common, but does occur in some places, e.g.
https://github.com/borglab/gtsam/blob/7f80c906c4cadc32e71acbe5209f2dba90ace779/tests/testExtendedKalmanFilter.cpp#L192-L194
or, strangely, `BoundingConstraint` redefines it:
https://github.com/borglab/gtsam/blob/7f80c906c4cadc32e71acbe5209f2dba90ace779/gtsam/slam/BoundingConstraint.h#L99-L102

But also, backwards compatibility?  What if users use it?

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h:593]: If you mean can I explain more in detail here, yes:

In c++, a derived class can refer to its base class either _with_ or _without_ template parameters.  This is a syntactical convenience: you can always refer to yourself (or your visible ancestors) from within your class scope.  See [injected-class-name](https://en.cppreference.com/w/cpp/language/injected-class-name) for full description.

Here's a simple example, where `Derived` can either use the `B` form _or_ the `B<T>` form.

```c++
template <typename T>
struct Base {
  using Type = T;
};

template <typename T>
struct Derived : public Base<T>{
  using Type1 = Base::Type;
  using Type2 = Base::<T>::Type;
};
```

As a note, this means instead of the common pattern
```c++
using Base = NoiseModelFactor4<T1, T2, T3, T4>;
```
I'm pretty sure we could instead just do
```c++
using Base = NoiseModelFactor4;
```

The reason this is problematic is because if you do
```c++
template <typename T>
using Derived = Base<T>
```

Just a weird quirk that anyone who inherits from `Derived` does NOT get class name injection so they _must_ use `Base<T>` and _cannot_ use just `Base`.  I'm not sure why the spec was written this way.

In one test file in the repo, someone does use just `NoiseModelFactor6` in a derived class without specifying the template parameters.  We could just change the unit test, but then that would be in conflict with:
> has to be 100% compatible of course.

Also, if we try doing something tricky like

```c++
class NoiseModelFactorN<...> {
  ...
  using NoiseModelFactor4 = This;
};
```

Then the `NoiseModelFactor4<T>` syntax doesn't work in the derived class.  So basically I'm almost certain there isn't a way to be able to use both syntaxes: `NoiseModelFactor4` and `NoiseModelFactor4<T1, T2, T3, T4>` using only `using`.

An alternative is to define structs:

```c++
template <...>
struct NoiseModelFactor4 : public NoiseModelFactorN<...> {
 public:
  using NoiseModelFactorN<...>::NoiseModelFactorN;
  using This = NoiseModelFactor4<...>;
};
```

But I don't see this as being significantly better than the typedefs.

One advantage of structs is that we could then move the deprecated `keyX()` functions and `using X` into there.

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h:473]: done

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h:32]: done

File: gtsam/mainpage.dox
Comment (gchenfc) [reply gtsam/mainpage.dox]: done

**Comment by jlblancoc** [comment]
Might be doable with the `typename... Args` with clearer syntax? See e.g. #444 
Just a first guess, perhaps it's not possible in this case for some reason.

**Comment by gchenfc** [comment]
> Might be doable with the `typename... Args` with clearer syntax? See e.g. #444 Just a first guess, perhaps it's not possible in this case for some reason.

Thanks so much for the tip @jlblancoc!  Would you mind clarifying which aspects you were thinking to include this?  It's currently being used quite a bit here, e.g.
https://github.com/borglab/gtsam/blob/11fd8612269690c9d95eba4a80ac5db705906aa1/gtsam/nonlinear/NonlinearFactor.h#L404-L406
https://github.com/borglab/gtsam/blob/11fd8612269690c9d95eba4a80ac5db705906aa1/gtsam/nonlinear/NonlinearFactor.h#L519-L520

but I'm hung up on all the "custom" additions for backwards compatibility, e.g. `Key key1 const { return keys_[0]; }` and `typedef X2 VALUE2` which are the majority of what is making this so ugly at the moment - would absolutely love any ideas/suggestions for alternatives if you have any!

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h:407]: @dellaert please advise on recommended course of action.  Options:
1) keep current
2) typedef to `void` if the type should not apply to this set of templates (cleaner using std::conditional).  e.g. `using X1 = std::conditional< (1 <= N), VALUES[0], void >`
3) delete X1, X2, ...  since they're rarely used (backwards compatibility violation)
4) change `#define NoiseModelFactor4 NoiseModelFactorN` to `struct NoiseModelFactor4 : public NoiseModelFactorN` and move these custom typedefs into there
5) other

**Comment by jlblancoc** [comment]
@gchenfc Ah, very good job then, IMO you are on the right track!  :+1: 

> but I'm hung up on all the "custom" additions for backwards compatibility, e.g. `Key key1 const { return keys_[0]; }`

I'm afraid gtsam is so used, it needs that backwards compatibility, so it should be kept despite the "ugliness" ;-)

**Comment by gchenfc** [comment]
Cool, thanks for sharing + advice! much appreciated :)

**Comment by gchenfc** [comment]
Just for posterity, recapping a quick offline discussion with Frank:

We'll get rid of the template meta-programming needed for the key1() key2(), etc. functions and `typedef X1`, X2, etc. and that'll make things much easier to read (but we'll still have the template parameter packs etc. for other NoiseModelFactorN stuff e.g. evaluateError, obviously).  In their place, we'll just have e.g.
```c++
class NoiseModelFactor3 : public NoiseModelFactorN {
 public:
  using X1 = VALUE<0>;
  Key key1() const { return keys_[0]; };
  ...
};
```

This is similar to the state of the code at commit [ea7d6947](https://github.com/borglab/gtsam/pull/947/files/ea7d769476e897f786799c5f774846460e880e19).  For example, [NoiseModelFactor1](https://github.com/borglab/gtsam/pull/947/files/ea7d769476e897f786799c5f774846460e880e19#diff-34f9b717c856740c8ada0fe7c4b1f746e979ddfebb5fedb48f73dc72248007a8R501-R533)

I'll revert to this commit, clean up, add a unit test to make sure serialization still behaves the same way, then should be ready to go.

**Comment by dellaert** [comment]
ping

**Comment by dellaert** [comment]
CI seem to be failing still. Please re-request review when needed.

**Comment by gchenfc** [comment]
Re-requested review, Python CI should pass after #1250

**Comment by mattking-smith** [comment]
@gchenfc Will I be able to test this NoiseModelFactorN in `Matlab`? What is holding up the merge on this PR? I know me and my lab mate have both needed this in the past.

File: gtsam/nonlinear/NonlinearFactor.h
Comment (varunagrawal) [inline gtsam/nonlinear/NonlinearFactor.h:313]: Would you be opposed to renaming this from `X` to `TYPE` or `T`? I can envision a scenario where a user sees `X` and thinks it is the symbol shorthand.

From a readability perspective, I prefer `MyFactor::TYPE<1>`.

File: gtsam/nonlinear/NonlinearFactor.h
Comment (varunagrawal) [inline gtsam/nonlinear/NonlinearFactor.h:334]: This can be a one line docstring
```
/// Default Constructor for I/O
```

File: gtsam/nonlinear/NonlinearFactor.h
Comment (varunagrawal) [inline gtsam/nonlinear/NonlinearFactor.h:367]: This is neat!

File: gtsam/nonlinear/NonlinearFactor.h
Comment (varunagrawal) [inline gtsam/nonlinear/NonlinearFactor.h:511]: Add the `GTSAM_DEPRECATED` attribute to these classes?

```cpp
class GTSAM_DEPRECATED NoiseModelFactor2 ...
```

File: gtsam/nonlinear/NonlinearFactor.h
Comment (varunagrawal) [inline gtsam/nonlinear/NonlinearFactor.h:547]: Same here `GTSAM_DEPRECATED`

File: gtsam/nonlinear/NonlinearFactor.h
Comment (varunagrawal) [inline gtsam/nonlinear/NonlinearFactor.h:585]: Same here `GTSAM_DEPRECATED`

File: gtsam/nonlinear/NonlinearFactor.h
Comment (varunagrawal) [inline gtsam/nonlinear/NonlinearFactor.h:626]: Same here `GTSAM_DEPRECATED`

File: gtsam/nonlinear/NonlinearFactor.h
Comment (varunagrawal) [inline gtsam/nonlinear/NonlinearFactor.h:671]: Same here `GTSAM_DEPRECATED`

**Review by varunagrawal — APPROVED** [review]
@gchenfc this looks great except for some nitpicks. Can you please make the modifications and we can land this?

File: gtsam/nonlinear/NonlinearFactor.h
Comment (varunagrawal) [reply gtsam/nonlinear/NonlinearFactor.h:313]: Or even `MyFactor::type<1>` similar to `MyFactor::key<1>` later.

Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h:313]: I'm happy to change, e.g. to `T<#>`, but my logic here was to match the old API, which uses `X1`, `X2`, ...
I think this was probably done in the original version since the input variable (e.g. in unwhitenedError, evaluateError, etc.) is always named with `x`, so the type should be capital `X`.  I guess `x` is used to denote an "input" variable, as in f(x).

Can you please re-confirm whether you prefer this change?  Then I'll apply it or resolve convo.

**Comment by gchenfc** [comment]
Thanks @varunagrawal for the review!  I'll go through and add DEPRECATED - I think you created that after I had originally made this PR so I'm glad you reminded me of this!  Can you also confirm the other 2 items and, if you still want those changes, I'll add those changes too.



> @gchenfc Will I be able to test this NoiseModelFactorN in `Matlab`? What is holding up the merge on this PR? I know me and my lab mate have both needed this in the past.

I don't think this can directly be used in Matlab since `NoiseModelFactorN` is templated and you need to supply the types of all the variables during compile-time.  Actually, now that I think of it, I forget how e.g. NoiseModelFactor2 is even used in Matlab, since you can't define custom factors?  Because `NoiseModelFactor1`, ..., `NoiseModelFactorN` are primarily used as base classes for defining custom factors.  But it would be easier (less copy-paste) to define e.g. `NoiseModelFactor8<T1, T2, ..., T8>` if hypothetically you wanted it.  If it's just for linear, we can do that in a different PR.  If it's for nonlinear, I'm not sure how this would help in Matlab since custom factors need to be defined in C++ anyway (unless using Fan's CustomFactor class).
And I think the PR was previously just waiting for review, so thanks varun again :)

File: gtsam/nonlinear/NonlinearFactor.h
Comment (varunagrawal) [reply gtsam/nonlinear/NonlinearFactor.h:313]: If the old API had `X` then let's match it.

File: gtsam/nonlinear/NonlinearFactor.h
Comment (varunagrawal) [reply gtsam/nonlinear/NonlinearFactor.h:334]: I've generally seen that we use `///` for single line comments. We definitely don't follow the Google Style in this case. 😅

**Comment by varunagrawal** [comment]
`GTSAM_DEPRECATED` was already present even before I started working with GTSAM. @dellaert and I have been deprecating stuff as we revisit components for hybrid estimation (e.g. `EliminateForMPE`) which is how I know about it, so passing that knowledge on. :)

@mattking-smith we'll be able to write a C++ class with 8+ variables (which isn't doable in current `develop`) and then we can wrap that class. We can talk about it offline (especially since now that I better understand what you're doing).

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h]: done

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h]: done

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h]: done

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h]: done

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h]: done

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h]: done

**Comment by gchenfc** [comment]
Thanks!  Will merge once CI completes, assuming that github lets me; we may still need an approval from @dellaert  since he previously requested changes (that have since been fixed I think).

File: gtsam/nonlinear/NonlinearFactor.h
Comment (dellaert) [reply gtsam/nonlinear/NonlinearFactor.h:313]: I noticed the mucking in the tests with avoiding collisions with symbol_shorthand::X, so I think we should revisit. I propose to rename

`VALUES -> ValueTypes` 

```
using ValueType = ...
```

File: gtsam/nonlinear/NonlinearFactor.h
Comment (dellaert) [inline gtsam/nonlinear/NonlinearFactor.h:322]: OptionalMatrix, and should it be `int i` rather than `typename T` ?

File: gtsam/nonlinear/NonlinearFactor.h
Comment (dellaert) [inline gtsam/nonlinear/NonlinearFactor.h:327]: KeyType, and same comment.

File: gtsam/nonlinear/NonlinearFactor.h
Comment (dellaert) [inline gtsam/nonlinear/NonlinearFactor.h:348]: why?

**Review by dellaert — COMMENTED** [review]
Some comments. This PR is a year old now :-) Could we meet and put a pin in it ? I really like the idea and want to have it in 4.2.

**Comment by dellaert** [comment]
@gchenfc

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h:322]: Changed the name, but I think it should still be templated on `typename T` instead of `int i` because it needs to accept a template parameter of type `ValueType`.

The reasoning I think is best described by cppreference description of [std::void_t](https://en.cppreference.com/w/cpp/types/void_t):
> Utility metafunction that maps a [sequence] of any types to the type void

Except in this case we're mapping any `ValueType` (non-sequence) to a `boost::optional<Matrix&>` type.  You may be able to imagine why this is helpful in `evaluateError`, since the arguments come in pairs of `(ValueType x1, ..., Matrix H1, ...)`.

In anticipation of the next question, `OptionalJacobian<N, gtsam::traits<ValueType>::N>` or something along those lines had issues with inheritance & backwards compatibility, IIRC.  But perhaps this would actually be possible in a future PR.

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h:327]: change the name, see comment above

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h]: Actually I'm not sure.  I changed it to remove this requirement.

No tests break when N = 1 and I added unit tests to also test the N=1 case so I think I might have just gotten overzealous with SFINAE at the time I was writing this.

File: gtsam/nonlinear/NonlinearFactor.h
Comment (gchenfc) [reply gtsam/nonlinear/NonlinearFactor.h]: Done.

**Comment by dellaert** [comment]
PS, re-reading some above, we definitely need to check that both wrappers (python and matlab) still work. @mattking-smith might be able to help with latter...

**Comment by gchenfc** [comment]
> PS, re-reading some above, we definitely need to check that both wrappers (python and matlab) still work. @mattking-smith might be able to help with latter...

Tested both and all pass!

Python is tested by CI and I also ran on my machine, all pass.

Matlab I tested on my machine and all pass, after needing to change some instances of `symbol` -> `Symbol` due to my computer having case insensitive filenames.

**Review by dellaert — APPROVED** [review]
This is an amazing piece of work, Gerry! Will merge so it’s ready to go into 4.2 :-)

**Pull request merged**
